index.js

```diff
@@ -646,28 +1125,85 @@
     );
   };
 
-  var invalidReturnValue = function(name, types, value) {
+  //  badTypeVar :: (Number, String) -> Node
+  var badTypeVar = function(index, typeVarName) {
+    return isNaN(index) ? findRetTypeVar(typeVarName) :
+                          findArgTypeVar(index, typeVarName);
+  };
+
+  var showValueAndType = function(typeFn, value) {
+    return show(value) + ' :: ' + typeFn([value]).join(', ');
+  };
+  
+  var showValuesAndTypes = function(typeFn, values) {
+    values = map(values, function(val) { return showValueAndType(typeFn, val); });
+    if (values.length > 4) {
+      values = values.slice(0, 2).concat('...').concat(values.slice(-2));
+    }
+    return values.join('\n    ');
+  };
+
+  //  showErrorValue :: (Integer, (a -> Type), a) -> String
+  var showErrorValue = function(num, typeFn, value) {
+    return (
+      num + ')  ' +
+      (isCollection(value) ? showValuesAndTypes(typeFn, value) :
+                             showValueAndType(typeFn, value)) +
+      '\n\n'
+    );
+  };
+
+  var conflictingTypeVar = function(sig, typeVarName, typeFn,
+                                    prevIndex, prevValue,
+                                    badIndex, badValue) {
+    if (prevIndex > badIndex) {
+      var tmpIndex = prevIndex; prevIndex = badIndex; badIndex = tmpIndex;
+      var tmpValue = prevValue; prevValue = badValue; badValue = tmpValue;
+    }
     return new TypeError(
-      LEFT_SINGLE_QUOTATION_MARK + name + RIGHT_SINGLE_QUOTATION_MARK +
-      ' is expected to return a value of type ' + types.join(' or ') +
-      '; returned ' + show(value)
+      'Type-variable constraint violation\n\n' +
+      showSignature(sig) + '\n' +
+      sigHighlights(sig, [
+        [true, findArgTypeVar(prevIndex, typeVarName)],
+        [true, badTypeVar(badIndex, typeVarName)]
+      ]) + '\n\n' +
+      showErrorValue(1, typeFn, prevValue) +
+      showErrorValue(2, typeFn, badValue) +
+      'Since there is no type of which all the above values are members, ' +
+      'the type-variable constraint has been violated.\n'
     );
   };
 
-  var invalidValue = function(name, types, value, index) {
-    return isNaN(index) ?
-      invalidReturnValue(name, types, value) :
-      invalidArgument(name, types, value, index);
+  var invalidValue = function(sig, expTypes, typeFn, index, value) {
+    return new TypeError(
+      'Invalid value\n\n' +
+      showSignature(sig) + '\n' +
+      sigHighlights(sig, [
+        [true, findVal(index)]
+      ]) + '\n\n' +
+      showErrorValue(1, typeFn, value) +
+      'The value at position 1 is not a member of ' +
+      expTypes.join(', ') + '.\n'
```

Comment:
I think `' or '` would make a better separator.

    The value at position 1 is not a member of Number or Integer

We could be fancy and use the `A`, `A or B`, `A, B, or C` format:

```javascript
//  oxford :: [Any] -> String
var oxford = function(xs) {
  return xs.length <= 2 ?
    xs.join(' or ') :
    xs.slice(0, -1).join(', ') + ', or ' + xs[xs.length - 1];
};
```